packages/x-data-grid-pro/src/hooks/features/serverSideLazyLoader/useGridDataSourceLazyLoader.ts

```diff
@@ -0,0 +1,410 @@
+import * as React from 'react';
+import { throttle } from '@mui/x-internals/throttle';
+import {
+  useGridApiEventHandler,
+  useGridSelector,
+  gridSortModelSelector,
+  gridFilterModelSelector,
+  GridEventListener,
+  GRID_ROOT_GROUP_ID,
+  GridGroupNode,
+  GridSkeletonRowNode,
+  gridPaginationModelSelector,
+  gridDimensionsSelector,
+} from '@mui/x-data-grid';
+import {
+  getVisibleRows,
+  GridGetRowsParams,
+  gridRenderContextSelector,
+} from '@mui/x-data-grid/internals';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import { findSkeletonRowsSection } from '../lazyLoader/utils';
+import { GRID_SKELETON_ROW_ROOT_ID } from '../lazyLoader/useGridLazyLoaderPreProcessors';
+
+enum LoadingTrigger {
+  VIEWPORT,
+  SCROLL_END,
+}
+
+const INTERVAL_CACHE_INITIAL_STATE = {
+  firstRowToRender: 0,
+  lastRowToRender: 0,
+};
+
+const getSkeletonRowId = (index: number) => `${GRID_SKELETON_ROW_ROOT_ID}-${index}`;
+
+/**
+ * @requires useGridRows (state)
+ * @requires useGridPagination (state)
+ * @requires useGridDimensions (method) - can be after
+ * @requires useGridScroll (method
+ */
+export const useGridDataSourceLazyLoader = (
+  privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    | 'pagination'
+    | 'paginationMode'
+    | 'unstable_dataSource'
+    | 'lazyLoading'
+    | 'lazyLoadingRequestThrottleMs'
+    | 'scrollEndThreshold'
+  >,
+): void => {
+  const sortModel = useGridSelector(privateApiRef, gridSortModelSelector);
+  const filterModel = useGridSelector(privateApiRef, gridFilterModelSelector);
+  const paginationModel = useGridSelector(privateApiRef, gridPaginationModelSelector);
+  const dimensions = useGridSelector(privateApiRef, gridDimensionsSelector);
+  const renderContext = useGridSelector(privateApiRef, gridRenderContextSelector);
+  const renderedRowsIntervalCache = React.useRef(INTERVAL_CACHE_INITIAL_STATE);
+  const previousLastRowIndex = React.useRef(0);
+  const loadingTrigger = React.useRef<LoadingTrigger | null>(null);
+  const isDisabled = !props.unstable_dataSource || props.lazyLoading !== true;
+
+  const heights = React.useMemo(
```

Comment:
Nitpick: drop unnecessary memoization and use `dimensions` directly where needed instead of `heights`?